core/src/main/java/org/apache/spark/shuffle/sort/ShuffleInMemorySorter.java

```diff
@@ -69,7 +72,11 @@ public int numRecords() {
     return pos;
   }
 
-  public void reset() {
+  public void freeMemoryAndResetPointerArray() {
+    if (consumer != null) {
+      consumer.freeArray(array);
+      this.array = consumer.allocateArray(initialSize);
```

Comment:
@davies - Thanks for your input. Even if another task grabs all the memory freed by the `consumer.freeArray(array)`,  `consumer.allocateArray(initialSize)` should always succeed because the `MemoryManager` guarantees that each task is allocated atleast 1/2N of the total shuffle memory, right? 

Manually releasing the array, bypassing the `Memorymanager` does not sound like a very clean approach to me. Also how would I do this `consumer.allocateArray(initialSize)` bypassing the memory manager?